### PR TITLE
[ENH] for rc3, Graph clean up, faster compilation, conv.assert_shape Theano flags, add unsafe in useless opt

### PR DESCRIPTION
**Pull request by nouiz**
I think we should include this in our rc3

### CODE REVIEW

File: theano/configdefaults.py [hunk 1]
```diff
@@ -127,6 +127,13 @@ def __str__(self):
     in_c_key=False)
 
 AddConfigVar(
+    'conv.assert_shape',
+    "If False, AbstractConv* ops won't add assert that verify that"
+    " the user provided shapes are also the one at run time",
+    BoolParam(True),
```
Comment (abergeron) [inline theano/configdefaults.py]: I would default this to False, since it seems like a debugging option.

**Comment by nouiz** [comment]
I agree with @abergeron. If @lamblin is OK with that, I'll include that in
rc3.

**Comment by nouiz** [comment]
closing and reopening to trigger travis.

File: theano/gof/opt.py [hunk 2]
```diff
@@ -607,11 +607,20 @@ def process_node(self, fgraph, node):
         # properly.
         # The clients should at least contain `node` itself!
         if node.inputs:
-            assert len(node.inputs[0].clients) > 0
-            assert (node, 0) in node.inputs[0].clients
+            # Take the smallest clients list. Some ops like elemwise
+            # have optimization that put constant as the first inputs.
+            # As constant have in general more clients then other type of nodes
```
Comment (Sentient07) [inline theano/gof/opt.py]: Sorry for a late comment, a small correction, here it should be "than" and not "then".

Comment (nouiz) [inline theano/gof/opt.py]: done thanks.

**Pull request merged**